machinist/mnode/mnode.go

```diff
@@ -67,3 +77,90 @@ func (n *Node) BeginPolling() error {
 		}
 	}
 }
+
+// TODO(adam): perform rollbacks if enroll fails
+func (n *Node) Enroll() error {
+	if os.Geteuid() != 0 && n.config.RequireRoot {
+		return errors.New("this command must be run as root since it touches the /etc/ssh directory")
+	}
+	pubKey, privKey, err := kcerts.GenerateED25519()
+	if err != nil {
+		return err
+	}
+	hcr := &auth.HostCertificateRequest{
+		Hostcert: pem.EncodeToMemory(&pem.Block{Type: "PUBLIC KEY", Bytes: ssh.MarshalAuthorizedKey(pubKey)}),
+		Hosts:    n.config.SSHPrincipals,
+	}
+	resp, err := n.AuthClient.HostCertificate(context.Background(), hcr)
+	if err != nil {
+		return err
+	}
+	if !n.config.ReWriteConfigs {
+		if err = anyFileExist(
+			n.config.CaPublicKeyLocation,
+			n.config.HostKeyLocation, n.config.HostCertificate()); err != nil {
+			return fmt.Errorf("rewriting configs are disabled, failed with err(s): %v", err)
+		}
+	}
+
+	enrollConfig := n.config.enrollConfigs
+	// Pam Installer Steps
+	n.Log.Infof("Executing Pam installation steps")
+	if err := InstallLibPam(n.Log); err != nil {
+		return err
+	}
+	if err := InstallPamSSHDFile(enrollConfig.PamSSHDLocation, n.Log); err != nil {
+		return err
+	}
+	if err := InstallPamScript(enrollConfig.PamSecurityLocation, n.Log); err != nil {
+		return err
+	}
+
+	//// Nss AutoUser Setup
+	if err := InstallNssAutoUserConf(n.config.LibNssConfLocation, n.config.NssConfig()); err != nil {
+		return err
+	}
+	if err := InstallNssAutoUser(n.Log); err != nil {
+		return err
+	}
+
+	// SSHD installer steps
+	if err := os.MkdirAll(filepath.Dir(n.config.SSHDConfigurationLocation), os.ModePerm); err != nil {
+		return err
+	}
+	sshdConfigContent, err := ReadSSHDContent(n.config.CaPublicKeyLocation, n.config.HostKeyLocation, n.config.HostCertificate())
+	if err != nil {
+		return err
+	}
+	n.Log.Infof("Writing SSHD Configuration")
+	if err := ioutil.WriteFile(n.config.SSHDConfigurationLocation, sshdConfigContent, 0644); err != nil {
+		return err
+	}
+	n.Log.Infof("Writing CA Public Key Configuration")
+	if err := ioutil.WriteFile(n.config.CaPublicKeyLocation, resp.Capublickey, 0644); err != nil {
+		return err
+	}
+	n.Log.Infof("Writing Host Cert")
+	if err := ioutil.WriteFile(n.config.HostCertificate(), resp.Signedhostcert, 0644); err != nil {
+		return err
+	}
+	n.Log.Infof("Writing Host Key")
+	pemBytes, err := privKey.SSHPemEncode()
+	if err != nil {
+		return err
+	}
+	if err := ioutil.WriteFile(n.config.HostKeyLocation, pemBytes, 0644); err != nil {
+		return err
+	}
+	return nil
+}
+
+func anyFileExist(names ...string) error {
+	var errs []error
+	for _, name := range names {
+		if _, err := os.Stat(name); err != nil {
```

Comment:
os.IsNotExist is supposed to be ignored. This code likely no longer works on a clean machine? Test it?